File: packages/control-plane/src/tenancy/resolver.ts
/**
 * TenantResolverChain — resolves tenant identity from request context.
 *
 * Chain of responsibility: tries each resolver in order, first match wins.
 * Results are cached per-request to avoid repeated DB lookups.
 */

import type { TenantResolver, RequestContext } from '@clawdesk/shared';

export class TenantResolverChain {
  private _resolvers: TenantResolver[];

  constructor(resolvers: TenantResolver[]) {
    this._resolvers = resolvers;
  }

  /**
   * Resolve tenant ID from request context.
   * Tries each resolver in order. First non-null result wins.
   * Returns null if no resolver can identify the tenant.
   */
  async resolve(context: RequestContext): Promise<string | null> {
    for (const resolver of this._resolvers) {
      const tenantId = await resolver.resolve(context);
      if (tenantId !== null) {
        return tenantId;
      }
    }
    return null;
  }

  /** Get the list of resolver names (for debugging/logging). */
  get resolverNames(): string[] {
    return this._resolvers.map((r) => r.name);
  }
}

// ─── Built-in Resolvers ──────────────────────────────────

/**
 * Resolves tenant from a request header (e.g., X-Tenant-ID or X-API-Key).
 */
export class HeaderTenantResolver implements TenantResolver {
  readonly name = 'header';
  private _headerName: string;
  private _lookupFn: (value: string) => Promise<string | null>;

  constructor(
    headerName: string,
    lookupFn: (headerValue: string) => Promise<string | null>,
  ) {
    this._headerName = headerName.toLowerCase();
    this._lookupFn = lookupFn;
  }

  async resolve(context: RequestContext): Promise<string | null> {
    const value = context.headers[this._headerName];
    if (!value) return null;
    return this._lookupFn(value);
  }
}

/**
 * Resolves tenant from subdomain (e.g., acme.clawdesk.com → "acme").
 */
export class SubdomainTenantResolver implements TenantResolver {
  readonly name = 'subdomain';
  private _rootDomain: string;
  private _lookupFn: (slug: string) => Promise<string | null>;

  constructor(
    rootDomain: string,
    lookupFn: (slug: string) => Promise<string | null>,
  ) {
    this._rootDomain = rootDomain.toLowerCase();
    this._lookupFn = lookupFn;
  }

  async resolve(context: RequestContext): Promise<string | null> {
    const hostname = context.hostname?.toLowerCase();
    if (!hostname) return null;

    // Skip if it's the root domain itself
    if (hostname === this._rootDomain || hostname === `www.${this._rootDomain}`) {
      return null;
    }

    // Skip IP addresses
    if (/^\d+\.\d+\.\d+\.\d+$/.test(hostname)) {
      return null;
    }

    // Extract subdomain
    if (!hostname.endsWith(`.${this._rootDomain}`)) {
      return null;
    }

    const subdomain = hostname.replace(`.${this._rootDomain}`, '');
    if (!subdomain || subdomain.includes('.')) {
      return null; // Skip multi-level subdomains
    }

    return this._lookupFn(subdomain);
  }
}

/**
 * Resolves tenant from JWT claims (e.g., tenant_id or org_id claim).
 */
export class JwtClaimTenantResolver implements TenantResolver {
  readonly name = 'jwt-claim';
  private _claimName: string;

  constructor(claimName: string = 'tenant_id') {
    this._claimName = claimName;
  }

  async resolve(context: RequestContext): Promise<string | null> {
    const value = context.claims?.[this._claimName];
    if (typeof value === 'string' && value.length > 0) {
      return value;
    }
    return null;
  }
}

File: packages/control-plane/clawdesk/tenancy/resolver.py
"""
TenantResolverChain — resolves tenant identity from request context.

Chain of responsibility: tries each resolver in order, first match wins.
"""

from __future__ import annotations

import re
from typing import Awaitable, Callable, Protocol

from clawdesk.models import RequestContext


class TenantResolver(Protocol):
    """Interface for tenant resolvers."""

    @property
    def name(self) -> str: ...

    async def resolve(self, context: RequestContext) -> str | None: ...


class TenantResolverChain:
    def __init__(self, resolvers: list[TenantResolver]) -> None:
        self._resolvers = resolvers

    async def resolve(self, context: RequestContext) -> str | None:
        """Try each resolver in order. First non-None result wins."""
        for resolver in self._resolvers:
            tenant_id = await resolver.resolve(context)
            if tenant_id is not None:
                return tenant_id
        return None

    @property
    def resolver_names(self) -> list[str]:
        return [r.name for r in self._resolvers]


# ─── Built-in Resolvers ──────────────────────────────────


LookupFn = Callable[[str], Awaitable[str | None]]

_IP_RE = re.compile(r"^\d+\.\d+\.\d+\.\d+$")


class HeaderTenantResolver:
    """Resolves tenant from a request header (e.g., X-Tenant-ID or X-API-Key)."""

    def __init__(self, header_name: str, lookup_fn: LookupFn) -> None:
        self._header_name = header_name.lower()
        self._lookup_fn = lookup_fn

    @property
    def name(self) -> str:
        return "header"

    async def resolve(self, context: RequestContext) -> str | None:
        value = context.headers.get(self._header_name)
        if not value:
            return None
        return await self._lookup_fn(value)


class SubdomainTenantResolver:
    """Resolves tenant from subdomain (e.g., acme.clawdesk.com → 'acme')."""

    def __init__(self, root_domain: str, lookup_fn: LookupFn) -> None:
        self._root_domain = root_domain.lower()
        self._lookup_fn = lookup_fn

    @property
    def name(self) -> str:
        return "subdomain"

    async def resolve(self, context: RequestContext) -> str | None:
        hostname = (context.hostname or "").lower()
        if not hostname:
            return None

        # Skip root domain itself
        if hostname == self._root_domain or hostname == f"www.{self._root_domain}":
            return None

        # Skip IP addresses
        if _IP_RE.match(hostname):
            return None

        # Extract subdomain
        suffix = f".{self._root_domain}"
        if not hostname.endswith(suffix):
            return None

        subdomain = hostname[: -len(suffix)]
        if not subdomain or "." in subdomain:
            return None  # Skip multi-level subdomains

        return await self._lookup_fn(subdomain)


class JwtClaimTenantResolver:
    """Resolves tenant from JWT claims (e.g., tenant_id or org_id claim)."""

    def __init__(self, claim_name: str = "tenant_id") -> None:
        self._claim_name = claim_name

    @property
    def name(self) -> str:
        return "jwt-claim"

    async def resolve(self, context: RequestContext) -> str | None:
        value = context.claims.get(self._claim_name)
        if isinstance(value, str) and value:
            return value
        return None
